planner/core/collect_column_stats_usage_test.go

```diff
@@ -12,210 +12,302 @@
+func getColumnName(c *C, is infoschema.InfoSchema, tblColID model.TableColumnID, comment CommentInterface) (string, bool) {
+	var tblInfo *model.TableInfo
+	var prefix string
+	if tbl, ok := is.TableByID(tblColID.TableID); ok {
+		tblInfo = tbl.Meta()
+		prefix = tblInfo.Name.L + "."
+	} else {
+		db, exists := is.SchemaByName(model.NewCIStr("test"))
+		c.Assert(exists, IsTrue, comment)
+		for _, tbl := range db.Tables {
+			pi := tbl.GetPartitionInfo()
+			if pi == nil {
+				continue
+			}
+			for _, def := range pi.Definitions {
+				if def.ID == tblColID.TableID {
+					tblInfo = tbl
+					prefix = tbl.Name.L + "." + def.Name.L + "."
+					break
+				}
+			}
+			if tblInfo != nil {
+				break
+			}
+		}
+		if tblInfo == nil {
+			return "", false
+		}
+	}
+	for _, col := range tblInfo.Columns {
+		if tblColID.ColumnID == col.ID {
+			return prefix + col.Name.L, true
+		}
+	}
+	return "", false
+}
 
+func checkColumnStatsUsage(c *C, is infoschema.InfoSchema, lp LogicalPlan, onlyHistNeeded bool, expected []string, comment CommentInterface) {
+	var tblColIDs []model.TableColumnID
+	if onlyHistNeeded {
+		_, tblColIDs = CollectColumnStatsUsage(lp)
+	} else {
+		tblColIDs, _ = CollectColumnStatsUsage(lp)
+	}
+	cols := make([]string, 0, len(tblColIDs))
+	for _, tblColID := range tblColIDs {
+		col, ok := getColumnName(c, is, tblColID, comment)
+		c.Assert(ok, IsTrue, comment)
+		cols = append(cols, col)
+	}
+	sort.Strings(cols)
+	c.Assert(cols, DeepEquals, expected, comment)
+}
+
+func (s *testPlanSuite) TestCollectPredicateColumns(c *C) {
+	defer testleak.AfterTest(c)()
 	tests := []struct {
-		sql string
-		res []string
+		pruneMode string
+		sql       string
+		res       []string
 	}{
 		{
 			// DataSource
-			sql: "select * from t1 where a > 2",
-			res: []string{"t1.a"},
+			sql: "select * from t where a > 2",
+			res: []string{"t.a"},
 		},
 		{
 			// DataSource
-			sql: "select * from t1 where b in (2, 5) or c = 5",
-			res: []string{"t1.b", "t1.c"},
+			sql: "select * from t where b in (2, 5) or c = 5",
+			res: []string{"t.b", "t.c"},
 		},
 		{
 			// LogicalProjection
-			sql: "select * from (select a + b as ab, c from t1) as tmp where ab > 4",
-			res: []string{"t1.a", "t1.b"},
+			sql: "select * from (select a + b as ab, c from t) as tmp where ab > 4",
+			res: []string{"t.a", "t.b"},
 		},
 		{
 			// LogicalAggregation
-			sql: "select b, count(*) from t1 group by b",
-			res: []string{"t1.b"},
+			sql: "select b, count(*) from t group by b",
+			res: []string{"t.b"},
 		},
 		{
 			// LogicalAggregation
-			sql: "select b, sum(a) from t1 group by b having sum(a) > 3",
-			res: []string{"t1.a", "t1.b"},
+			sql: "select b, sum(a) from t group by b having sum(a) > 3",
+			res: []string{"t.a", "t.b"},
 		},
 		{
 			// LogicalAggregation
-			sql: "select count(*), sum(a), sum(c) from t1",
+			sql: "select count(*), sum(a), sum(c) from t",
 			res: []string{},
 		},
 		{
 			// LogicalAggregation
```

Comment:
Wrong comment

Reply:
`select count(*), sum(a), sum(c) from t` has no `LogicalUnionAll`.

Reply:
I mean lines 120-122

Reply:
`union` is the same as `union distinct`, which eliminates duplicate elements. We use `LogicalAggregation` to implement it. So we don't have `LogicalUnionAll` here.